File: Day-4/Leader/Leader.py
def find_leaders(arr):
    leaders = []
    max_from_right = float('-inf')

    for i in range(len(arr) - 1, -1, -1):
        if arr[i] > max_from_right:
            leaders.append(arr[i])
            max_from_right = arr[i]

    return leaders[::-1]  # Reverse to maintain the original order
if __name__ == "__main__":
    arr = [16, 17, 4, 3, 5, 2]
    print(find_leaders(arr))  # Output: [17, 5, 2]
# Time Complexity: O(n) - We traverse the array once to find the leaders.
# Space Complexity: O(k) - Where k is the number of leaders found, which in the worst case can be O(n).

File: Day-4/Leader/Leader.c
#include<stdio.h>
void leader(int arr[],int n)
{
    int i,j,flag;
    printf("Leaders in the array: ");
    for(i=0;i<n;i++)
    {
        flag=0;
        for(j=i+1;j<n;j++)
        {
            if(arr[i]<arr[j])
            {
                flag=1;
                break;
            }
        }
        if(flag==0)
        {
            printf("%d ",arr[i]);
        }
    }
}
int main()
{
    int n,i;
    printf("Enter the size of the array: ");
    scanf("%d",&n);
    int arr[n];
    printf("Enter the elements of the array: ");
    for(i=0;i<n;i++)
    {
        scanf("%d",&arr[i]);
    }
    leader(arr,n);
    return 0;
}
// Time Complexity: O(n^2)
// Space Complexity: O(1)